libraries/lib-mixer/Mix.cpp

```diff
@@ -78,6 +96,214 @@ size_t FindBufferSize(
+            channelFlags[1] = 1;
+         else
+            channelFlags[0] = 1;
+      }
+   }
+   bool CanMakeMono() const override { return mpMap == nullptr; }
+};
+
+class StereoToMono final : public DownmixSource
+{
+   AudioGraph::Source& mMasterSource;
+public:
+
+   StereoToMono(AudioGraph::Source& masterSource)
+      : mMasterSource(masterSource)
+   { }
+
+   AudioGraph::Source& GetDownstream() const override { return mMasterSource; }
+   size_t NChannels() const override { return 2; }
+   float GetChannelGain(size_t) const override { return 1.f; }
+   void FindChannelFlags(unsigned char* channelFlags, size_t numChannels, size_t) override
+   {
+      for(size_t i = 0; i < numChannels; ++i)
+         channelFlags[i] = true;
+   }
+   bool CanMakeMono() const override { return true; }
+};
+
+//! Combines multiple audio graph sources into a single source
+class DownmixStage final : public AudioGraph::Source
+{
+   std::vector<std::unique_ptr<DownmixSource>> mDownmixSources;
+   // Resample into these buffers, or produce directly when not resampling
+   AudioGraph::Buffers mFloatBuffers;
+   size_t mNumChannels;
+   Mixer::ApplyGain mApplyGain;
+
+public:
+
+   DownmixStage(std::vector<std::unique_ptr<DownmixSource>> downmixSources,
+                size_t numChannels,
+                size_t bufferSize,
+                Mixer::ApplyGain applyGain)
+      : mDownmixSources(std::move(downmixSources))
+    // PRL:  Bug2536: see other comments below for the last, padding argument
+    // TODO: more-than-two-channels
+    // Issue 3565 workaround:  allocate one extra buffer when applying a
+    // GVerb effect stage.  It is simply discarded
+    // See also issue 3854, when the number of out channels expected by the
+    // plug-in is yet larger
+      , mFloatBuffers { 3, bufferSize, 1, 1 }
+      , mNumChannels(numChannels)
+      , mApplyGain(applyGain)
+   {
+      
+   }
+
+   ~DownmixStage() override = default;
+
+   bool AcceptsBuffers(const Buffers& buffers) const override
+   {
+      return buffers.Channels() == mNumChannels &&
+         AcceptsBlockSize(buffers.BlockSize());  
+   }
+
+   bool AcceptsBlockSize(size_t blockSize) const override
+   {
+      return blockSize <= mFloatBuffers.BufferSize();
+   }
+
+   std::optional<size_t> Acquire(Buffers& data, size_t maxToProcess) override
+   {
+      // TODO: more-than-two-channels
+      auto maxChannels = std::max(2u, mFloatBuffers.Channels());
+      const auto channelFlags = stackAllocate(unsigned char, mNumChannels);
+      const auto gains = stackAllocate(float, mNumChannels);
+      if (mApplyGain == Mixer::ApplyGain::Discard)
+         std::fill(gains, gains + mNumChannels, 1.0f);
+
+      size_t maxOut = 0;
+      for (auto c = 0;c < data.Channels(); ++c)
+         data.ClearBuffer(c, maxToProcess);
+
+      for (auto& downmixSource : mDownmixSources)
+      {
+         auto oResult = downmixSource->GetDownstream().Acquire(mFloatBuffers, maxToProcess);
```

Comment:
I was also confused by that. We have a source, from which we'd normally like to acquire data, but to do this we have to ask its _downstream_ to provide us with data. Ruined me poor brain.
I think what might have brought you to this naming is this:
```cpp
      AudioGraph::Source *pDownstream = &source;
      for (const auto &stage : input.stages)
         if (
            auto& pNewDownstream =
               RegisterEffectStage(*pDownstream, sequence->NChannels(), stage, outRate))
         {
            pDownstream = pNewDownstream.get();
         }
      mixdownSources.emplace_back(
         std::make_unique<SequenceDownmixSource>(
            *pDownstream,
            *sequence,
            pMixerSpec ? &pMixerSpec->mMap[i] : nullptr
         )
      );
```
In that case, `pDownstream` is actually an upstream. It's just called that way because, at the end of this loop, it's supposed to point to the last source of the graph, or its "downstream". But from the point of view of a `DownmixSource`, it's an upstream.

I would "simply" remove the `GetDonwstream()` method and give the `DownmixSource` the `Acquire` , `Release` and `Remaining` methods.

Reply:
> I would "simply" remove the GetDonwstream() method and give the DownmixSource the Acquire , Release and Remaining methods.

I'd preferred not to do that because that will simply bloat the interface.

Reply:
> It's just called that way because, at the end of this loop, it's supposed to point to the last source of the graph, or its "downstream

And it is a downstream of individual processing sequence from the point of view of `DownmixStage`.